### DISCUSSION TITLE
Discussion on How to Move Forward with Blender 4.0 Material Changes

### DISCUSSION THREAD

**Discussion by StandingPadAnimations** [Moo-Ack-Productions/MCprep#500] [General Discussions]
As Blender 4.0 is getting closer to release, we now need to start looking into how we can move forward with material changes. While MCprep 3.5 was tested with Blender 4.0, the Blender Foundation introduced Principled V2 after we release MCprep 3.5, which makes MCprep's prep material system broke in Blender 4.0 (#494). In addition, there's currently an idea on the table that would revamp the materials system in MCprep (#274). 

Thus, this gives us a good opportunity to discuss how we should move forward with MCprep's material system, among both us developers and the community as a whole. 

Currently, I have 2 options in mind (but these definitely aren't the only options we're limited to):
1. Get MCprep's existing system to work in Blender 4.0
    - Pros:
        - Least friction with users
        - Existing system is “generation based”, so we have full control over the node tree
        - Already has existing code for previous versions
        - Proven itself over several years
    - Cons
        - A lot of code, especially when compatibility is factored in
        - Not very controllable by the user. When I talk with Blender users who started with Minecraft rendering, the biggest complaint about MCprep I hear is that the material system takes away too much control, and the materials have complex node trees that aren't wanted (although this is mostly solved with the “Simple” option)
        - Breaks if Blender changes inputs
2. Introduce all or a subset of the system in [Vivy](https://github.com/StandingPadAnimations/Vivy) (I should preface this by saying I am slightly biased towards Vivy as I created it). [A demonstration cam be found here](https://mastodon.art/@standingpad/111327393975731655)
    - Pros:
        - Template based, so Blender will update the nodetrees automatically
        - Gives a ton of control to the user; it's all templates
        - Less likely to break
        - Allows reducing code size
    - Cons:
        - Tied to a blend file and JSON pair. While Vivy as an addon allows exporting the files as zip files (and importing said zip files), it's still an extra thing users have to keep track of
        - Complex: to handle as many cases as MCprep's current system handles, Vivy has a system of refinements, which can be hard to grasp for new users
        - Makes MCprep not quite the “install and done” solution that users are used to
        - Completely upends what users are used to if introduced
        - Not nearly as proven; while it has been tested and showcased in a couple of videos on the MCprep Discord server, it's not nearly as proven as the current system, which has been proven itself well over the past several years
  
Regardless of which option is chosen (or if we decide to go with an idea that results from this discussion), I think this rare opportunity will lead to some interesting conversation about what is perhaps the most fundamental part of MCprep, especially as this leads to other ideas, such as:
- Layering resource packs (#123)
- Improving PBR materials to be more accurate to how they're meant to behave (#431)
- Improving sync materials (whose code is actually the core of what Vivy uses)
- Water shaders (#163)
- Etc.

**Comment by StandingPadAnimations** [comment]
Fitting how this is #500

**Comment by CFeeney333** [comment]
I wonder about the possibility of maintaining an mcprep version for compatibility with blender 3.6 LTS - since it is an LTS, then integrate the revamped material system for a version compatible with blender 4, then maintain these two mcprep versions simultaneously. At least that will cut down on the code for keeping one version of mcprep compatible with all blender versions... do you reckon that's a feasible option?

**Reply by zNightlord** [reply]
That would mean having both exist, don't cut down on the code in the main root.
Split into 2 branch of options (that would be the process of packaging a ready downloadable zip or something else)

**Reply by StandingPadAnimations** [reply]
An LTS MCprep release could be an option, though that might be confusing for users (users are an interesting group of people, able to do things you would think isn't possible)

**Comment by StandingPadAnimations** [comment]
Closing this as Blender 4.0 is now supported
